// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import com.ctre.phoenix6.CANBus;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Transform2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.geometry.Translation3d;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.RobotBase;
import frc.robot.util.AllianceFlipped;
import frc.robot.util.AllianceFlippedUtil;
import edu.wpi.first.apriltag.AprilTagFieldLayout;
import edu.wpi.first.apriltag.AprilTagFields;
import frc.robot.util.AllianceFlipped;
import frc.robot.util.AllianceFlippedUtil;

/**
 * The Constants class provides a convenient place for teams to hold robot-wide numerical or boolean
 * constants. This class should not be used for any other purpose. All constants should be declared
 * globally (i.e. public static). Do not put anything functional in this class.
 *
 * <p>It is advised to statically import this class (or one of its inner classes) wherever the
 * constants are needed, to reduce verbosity.
 */
public final class Constants {
  public static final int DRIVE_CONTROLLER = 0;
  public static final int OPERATOR_CONTROLLER = 1;

  // "rio" is the default bus. Use "CANivoreName" if you have a CANivore.
  public static final CANBus kDriveCANBus = new CANBus("lil clanker");


  public static final int FRONT_LEFT_MODULE_DRIVE_MOTOR = 2; 
  public static final int FRONT_LEFT_MODULE_STEER_MOTOR = 1;
  public static final int FRONT_LEFT_MODULE_STEER_ENCODER = 9; 
  public static final double FRONT_LEFT_MAGNET_OFFSET = 0.109619;
  // -0.002197; 

  public static final int FRONT_RIGHT_MODULE_DRIVE_MOTOR = 4;  
  public static final int FRONT_RIGHT_MODULE_STEER_MOTOR = 3;
  public static final int FRONT_RIGHT_MODULE_STEER_ENCODER = 10; 
  public static final double FRONT_RIGHT_MAGNET_OFFSET = 0.222412;
  // -0.547852; 

  public static final int BACK_LEFT_MODULE_DRIVE_MOTOR = 8; 
  public static final int BACK_LEFT_MODULE_STEER_MOTOR = 7; 
  public static final int BACK_LEFT_MODULE_STEER_ENCODER = 12; 
  public static final double BACK_LEFT_MAGNET_OFFSET = 0.247559;
  // -0.314209; 

  public static final int BACK_RIGHT_MODULE_DRIVE_MOTOR = 6; 
  public static final int BACK_RIGHT_MODULE_STEER_MOTOR = 5; 
  public static final int BACK_RIGHT_MODULE_STEER_ENCODER = 11; 
  public static final double BACK_RIGHT_MAGNET_OFFSET = 0.149658;
  // -0.610840;  

  // Define the conventional order of our modules when putting them into arrays
  public static final int FRONT_LEFT =0;
  public static final int FRONT_RIGHT =1;
  public static final int REAR_LEFT =2;
  public static final int REAR_RIGHT =3;

  public static final boolean kFrontLeftDriveEncoderReversed = false;
  public static final boolean kFrontRightDriveEncoderReversed = false;
  public static final boolean kRearLeftDriveEncoderReversed = true;
  public static final boolean kRearRightDriveEncoderReversed = true;

  public static final boolean kFrontLeftTurningEncoderReversed = false;
  public static final boolean kFrontRightTurningEncoderReversed = false;
  public static final boolean kRearLeftTurningEncoderReversed = true;
  public static final boolean kRearRightTurningEncoderReversed = true;

  public static final double kWheelDiameterMeters = 0.1016; //0.098; // 0.09398; // 3.7 in

  // The drive encoder reports in RPM by default. Calculate the conversion factor
  // to make it report in meters per second.
  public static final double kDriveGearRatio = 6.03;
  public static final double kDriveConversionFactor = (kWheelDiameterMeters * Math.PI) / kDriveGearRatio;

  public static final double kTurningGearRatio = 287/11;

  public static final double kTurnPositionConversionFactor = 12.8;

  public static final double kMaxSpeedMetersPerSecond = 5;
  // Units are meters.
  // Distance between centers of right and left wheels on robot
  public static final double kTrackWidth = 0.51435;
  
  // Distance between front and back wheels on robot
  public static final double kWheelBase = 0.51435;

  public static final double ROT_HUB_ALIGNMENT_P = 0.005;

  public static final double ROT_TOLERANCE_HUB_ALIGNMENT = 0.05;

  public static final double ROT_SETPOINT_HUB_ALIGNMENT = 0;

	public static final double DONT_SEE_TAG_WAIT_TIME = 1;
	public static final double POSE_VALIDATION_TIME = 0.3;

  // Units are meters per second
  public static final double kMaxTranslationalVelocity = 6784 / 60.0 *
  (1/kDriveGearRatio) *
  kWheelDiameterMeters * Math.PI;

  // Units are radians per second
  public static final double kMaxRotationalVelocity = kMaxTranslationalVelocity /
  Math.hypot(kTrackWidth / 2.0, kWheelBase / 2.0);; //max 5.0

  //The locations f
  //*or the modules must be relative to the center of the robot. 
  // Positive x values represent moving toward the front of the robot 
  // Positive y values represent moving toward the left of the robot.
  public static final SwerveDriveKinematics kDriveKinematics =
  new SwerveDriveKinematics(
      new Translation2d(kWheelBase / 2.0, kTrackWidth / 2.0),   // front left
      new Translation2d(kWheelBase / 2.0, -kTrackWidth / 2.0),  // front right
      new Translation2d(-kWheelBase / 2.0, kTrackWidth / 2.0),  // rear left
      new Translation2d(-kWheelBase / 2.0, -kTrackWidth / 2.0)  // rear right
      );


      
  public static final boolean kGyroReversed = false;

  public static final double kDriveP = 0.004; //Usually 0.05, 0.000000000000001?
  public static final double kDriveI = 0.0;
  public static final double kDriveD = 1; //Usually 0.0, 0.9;?
  public static final double kDriveFF = 0.1;
  public static final double SLEW_RATE_LIMITER = 3;

  public static final double kTurningP = 0.7; //Usually 0.05
  public static final double kTurningI = 0.0;
  public static final double kTurningD = 0.06;
  public static final double kAcceleration = 4;

  public static final int RESET_NAVX_BUTTON = 8;

  public static final Mode simMode = Mode.REAL;
  public static final Mode currentMode = RobotBase.isReal() ? Mode.REAL : simMode;

  public static final double AIM_kP = 0.015; //kP for autoAim

  public static final double DRIVER_DEADBAND = 0.01; 
  public static final double TRIGGER_DEADBAND = 0.7; 


  public static enum Mode {
    /** Running on a real robot. */
    REAL,

    /** Running a physics simulator. */
    SIM,

    /** Replaying from a log file. */
    REPLAY
  }
    
    public static final int INTAKE_TRIGGER = 3;
    public static final int OUTTAKE_TRIGGER = 2;//RETURN TO OLD SETTING LATER

    //Intake Constants
    public static final int INTAKE_MOTOR = 17;
    public static final double INTAKE_SPEED = .65;
    public static final int INTAKE = 0;

   

  // //Limelight Constants
  //   public static final double ROT_SETPOINT_REEF_ALIGNMENT = 0.0;
  //   public static final double ROT_TOLERANCE_REEF_ALIGNMENT = 1.0;
    public static final double TRANSLATION_TOLERANCE = 0.03;
    public static final double TX_TOLARANCE = 0.5;
    // public static final double X_TOLERANCE_REEF_ALIGNMENT = 0.01;    
  //   public static final double Y_SETPOINT_RIGHT_REEF_ALIGNMENT = 0.0;
  //   public static final double Y_SETPOINT_LEFT_REEF_ALIGNMENT = 0.0;
  //   public static final double Y_TOLERANCE_REEF_ALIGNMENT = 0.01;

  //Belt Constants
    // public static final int BELT_MOTOR_ID = 18;

  //GroundIntake Constants
    public static final int PIVOT_MOTOR_ID = 18;
    public static final double PIVOT_DOWN_POSITION = 2;
    public static final double PIVOT_UP_POSITION = 0;
    public static final int GROUND_INTAKE_ID = 17;
    public static final double GROUND_PIVOT_SPEED = 0.15;
    public static final int PIVOT_ABS_ENC_DIO =4;
    public static final int SECOND_GROUND_INTAKE_ID = 22;

  //Belt Constants
    public static final int BELT_RIGHT_MOTOR_ID = 14;
    public static final int BELT_LEFT_MOTOR_ID = 19;
    //CanRange
    //public static final int CANRANGE_ID = 20;


    //Shooter constants
    public static final int LEFT_MOTOR_ID = 15;
    public static final int RIGHT_MOTOR_ID = 16;
    public static final int MIDDLE_MOTOR_ID = 20;

    //HOOD constants
    public static final int HOOD_MOTOR_ID = 13;
    //CANdle
    public static final int CANDLE_ID = 19;

    //Hub positions
    public static final double distanceToFrontOfHub = Units.inchesToMeters(158.6);
    public static final double widthOfHub = Units.inchesToMeters(47);
    public static final double xCoordOfHub = distanceToFrontOfHub + widthOfHub / 2.0;

    public static final double fieldWidth = Units.inchesToMeters(317.7);
    public static final double yCoordOfHub = fieldWidth / 2.0;

    public static final AllianceFlipped<Translation2d> HUB = AllianceFlippedUtil.fromBlue(new Translation2d(xCoordOfHub, yCoordOfHub));

    //Expanding hopper
    public static final int EXTENSION_MOTOR_ID = 21;


    //Passing
    public static final double xCoordOfPass = distanceToFrontOfHub / 2.0;

    //From DS POV
    public static final double yCoordOfLeftPass = fieldWidth * 3.0 / 4.0;
    public static final double yCoordOfRightPass = fieldWidth * 1.0 / 4.0;

    public static final AllianceFlipped<Translation2d> LeftPass = AllianceFlippedUtil.fromBlue(new Translation2d(xCoordOfPass, yCoordOfLeftPass));
    public static final AllianceFlipped<Translation2d> RightPass = AllianceFlippedUtil.fromBlue(new Translation2d(xCoordOfPass, yCoordOfRightPass));

} 
